1-Draft/RFCNNN-WindowsPS-Compatibility-Utilities.md

```diff
@@ -0,0 +1,106 @@
+
+## Motivation
+
+    As a PowerShell user,
+    I can use modules written for Windows PowerShell in PowerShell Core,
+    so that I could migrate from Windows PowerShell to PowerShell Core.
+
+## User Experience
+
+Example that shows using commands from 'WindowsPS-only' module in PS Core:
+```PowerShell
+PS C:\> $PSVersionTable.PSEdition
+Core
+PS C:\> Import-Module DesktopOnlyModule -AsWindowsCompatibility
+PS C:\> (Get-Module DesktopOnlyModule).CompatiblePSEditions
+Desktop
+PS C:\> DesktopOnlyModuleFunction
+Success
+```
+Proof-of-concept test was successfull.
+
+For reference, here is current behaviour that this RFC is targeting to change:
+```PowerShell
+PS C:\> $PSVersionTable.PSEdition
+Core
+PS C:\> Import-Module DesktopOnlyModule
+Import-Module : Module 'C:\windows\system32\WindowsPowerShell\v1.0\Modules\DesktopOnlyModule\DesktopOnlyModule.psd1' does not support current PowerShell edition 'Core'. Its supported editions are 'Desktop'. Use 'Import-Module -SkipEditionCheck' to ignore the compatibility of this module.
+```
+
+## Specification
+
+Today a module manifest (since Windows PowerShell 5) may contain a `CompatiblePSEditions` property to indicate if it is compatible with PowerShell Core (value of `Core`).
+Initial proposal was that absense of `Core` value in `CompatiblePSEditions` during module import would signal to PS Core that it should process the module as described in this RFC, however PS Core still allows to load non-`Core` modules without any warnings if they are not located under System32 module path. This means that `CompatiblePSEditions` property can not be used for deciding whether to load a module as WinPS module or not.
+Current plan is to add a new `-AsWindowsCompatibility` parameter for `Import-Module`.
+
+When Import-Module in PS Core detects that user is attempting to load a module with `-AsWindowsCompatibility` parameter:
+  1. it creates a hidden `WindowsPS Compatibility` Windows PS process (unless one already exists)
+  2. creates PS Remoting connection to its IPC pipe
+  3. loads the module in remote Windows PS process
+  4. generates local proxy module/commands using IPC remoting connection and `Import-PSSession` cmdlet
+  5. when 'WindowsPS-only' module unload request is detected - unload module in remote process, close PS remoting channel, close remote process
+
+### PS Remoting Transport
+
+IPC / Named Pipe is to be used for connections to remote PowerShell process. Today a named pipe listener is created with each Windows PowerShell process. IPC transport has good performance and secured endpoints, however (unlike, for example, SSH transport) it is not currently supported by New-PSSession.

```

Comment:
I think you should call out here how this is different from the WindowsCompatibility module.  Specifically it doesn't rely on WinRM remoting.

Reply:
Looks like this is covered in 'Alternate Proposals and Considerations'.